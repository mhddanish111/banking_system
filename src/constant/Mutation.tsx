import { gql } from '@apollo/client';

export const CREATE_USER = gql`
  mutation CreateUser($userInput: UserInput) {
    createUser(userInput: $userInput) {
      firstName
      lastName
      message
    }
  }
`;

export const CHANGE_PASSWORD = gql`
  mutation ChangePassword($changePasswordInput: ChangePasswordInput) {
    changePassword(changePasswordInput: $changePasswordInput) {
      message
      success
      data {
        firstName
      }
    }
  }
`;

export const OPEN_ACCOUNT = gql`
  mutation CreateAccount($accountInput: AccountInput) {
    createAccount(accountInput: $accountInput) {
      _id
      accountNumber
      amount
      accountType
      categoryName
      limit
      minAmount
      maxAmount
      accountTypeId
      categoryId
      userId
      message
      type
      debitCardNumber
      debitCardCSVNumber
      expiryYear
      expiryMonth
    }
  }
`;

export const DEPOSIT = gql`
  mutation Deposit($depositInput: DepositInput) {
    deposit(depositInput: $depositInput) {
      transactionResult {
        _id
      }
      accounts {
        _id
        type
        accountType
        categoryName
        accountNumber
        amount
        minAmount
        maxAmount
        limit
        charge
        debitCardNumber
        debitCardCSVNumber
        expiryYear
        expiryMonth
        users {
          _id
          firstName
          lastName
        }
      }
    }
  }
`;

export const WITHDRAWAL = gql`
  mutation Withdrawal($withdrawalInput: WithdrawalInput) {
    withdrawal(withdrawalInput: $withdrawalInput) {
      transactionResult {
        _id
      }
      accounts {
        _id
        type
        accountType
        categoryName
        accountNumber
        amount
        minAmount
        maxAmount
        limit
        charge
        debitCardNumber
        debitCardCSVNumber
        expiryYear
        expiryMonth
        users {
          _id
          firstName
          lastName
        }
      }
    }
  }
`;
