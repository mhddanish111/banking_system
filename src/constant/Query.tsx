import { gql } from '@apollo/client';

export const LOGIN_USER = gql`
  query Login($logInInput: LogInInput) {
    login(logInInput: $logInInput) {
      firstName
      lastName
      userId
      token
      isValid
      message
      userType
    }
  }
`;

export const USER_DETAILS = gql`
  query UserDetails {
    userDetails {
      _id
      firstName
      lastName
      email
      mobile
      occupation
      income
      aadhaarNumber
      panNumber
      govtId
      govtIdNumber
      address
      city
      state
      postalCode
      isActive
      message
    }
  }
`;

export const GET_ACCOUNT_TYPES = gql`
  query AccountTypes {
    accountType {
      _id
      name
      type
      isActive
      category {
        _id
        name
        benefit
        isActive
        minAmount
        maxAmount
        limit
        charge
      }
    }
  }
`;

export const GET_ACCOUNTS = gql`
  query Accounts {
    accounts {
      _id
      type
      accountType
      categoryName
      accountNumber
      amount
      minAmount
      maxAmount
      limit
      charge
      debitCardNumber
      debitCardCSVNumber
      expiryYear
      expiryMonth
      users {
        _id
        firstName
        lastName
      }
    }
  }
`;

export const GET_ACCOUNT_WITH_TYPE = gql`
  query AccountWithType($accountWithTypeInput: AccountWithTypeInput) {
    accountWithType(accountWithTypeInput: $accountWithTypeInput) {
      _id
      type
      accountType
      categoryName
      accountNumber
      amount
      minAmount
      maxAmount
      limit
      charge
      debitCardNumber
      debitCardCSVNumber
      expiryYear
      expiryMonth
      users {
        _id
        firstName
        lastName
        aadhaarNumber
        panNumber
      }
    }
  }
`;

export const GET_TRANSACTION = gql`
  query Transaction($transactionInput: TransactionInput) {
    transaction(transactionInput: $transactionInput) {
      count {
        offset
        limit
        totalCount
      }
      result {
        _id
        transactionId
        accountNumber
        amount
        withdrawalAmount
        depositAmount
        transactionDate
        accountType
        categoryName
        transactionType
      }
    }
  }
`;
