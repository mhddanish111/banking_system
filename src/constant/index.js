export const USER_KEY = 'user_key';

export const AccountTypeList = [
  { label: 'Accounts', value: 'Accounts' },
  { label: 'Cards', value: 'Cards' },
];
