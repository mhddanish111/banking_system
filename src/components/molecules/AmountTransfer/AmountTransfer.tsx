import React, { useState, useEffect } from 'react';
import { filter, get, isEmpty, find, trim, parseInt } from 'lodash';
import { useMutation } from '@apollo/client';
import { useForm } from 'react-hook-form';
import Heading from '../../atoms/Heading';
import SelectDropdown from '../../atoms/SelectDropdown';
import Button from '../../atoms/Button';
import ErrorMessage from '../../atoms/ErrorMessage';
import { AccountTypeList } from '../../../constant';
import { DEPOSIT } from '../../../constant/Mutation';
import { accountDropdownList } from '../../../utils';
import { addAccounts, addCards } from '../../../context/actions';
import GetContext from '../../../context/GetContext';

type FormData = {
  accountNumber: string;
  amount: string;
};

const AmountTransfer: React.FC<any> = (props) => {
  const { dispatch }: any = GetContext();
  const [submitDeposit] = useMutation(DEPOSIT);
  const [errorMessage, setErrorMessage] = useState<String>('');
  const [isSuccess, setIsSuccess] = useState<Boolean>(false);
  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<FormData>({ mode: 'onChange' });
  const {
    state,
    type,
    selectedId,
    defaultAccountNumber,
    setIsCallTransaction,
  } = props;
  const [accountState, setAccountState] = useState<any>({
    type: '',
    selectedAccountList: [],
    selectedAccount: {},
    accountDropdown: [],
    selectedAccountNumber: '',
    selectedAccountType: '',
    selectedId,
  });

  useEffect(() => {
    if (selectedId !== accountState.selectedId) {
      setAccountState({
        type: '',
        selectedAccountList: [],
        selectedAccount: {},
        accountDropdown: [],
        selectedAccountNumber: '',
        selectedAccountType: '',
        selectedId,
      });
    }
  }, [accountState, selectedId]);
  const handlerAccount = (e: any) => {
    e.preventDefault();
    let { accounts, cards } = state;
    const { value } = e.target;
    if (type === 'Accounts') {
      accounts = filter(accounts, (item) => item._id !== selectedId);
    }
    if (type === 'Cards') {
      cards = filter(cards, (cardItem) => cardItem._id !== selectedId);
    }
    const other = { label: 'other', value: 'other' };
    if (value === 'Accounts') {
      const list = accountDropdownList(accounts);
      list.push(other);
      const singleAccount = get(accounts, ['0'], {});
      setAccountState({
        type: value,
        selectedAccountList: accounts,
        selectedAccount: singleAccount,
        accountDropdown: list,
        selectedAccountNumber: singleAccount?.accountNumber || '',
        selectedAccountType: singleAccount?.categoryName || 'other',
        selectedId,
      });
    }
    if (value === 'Cards') {
      const singleCard = get(cards, ['0'], {});
      const list = accountDropdownList(cards);
      list.push(other);
      setAccountState({
        type: value,
        selectedAccountList: cards,
        selectedAccount: singleCard,
        accountDropdown: list,
        selectedAccountNumber: singleCard?.accountNumber || '',
        selectedAccountType: singleCard?.categoryName || 'other',
        selectedId,
      });
    }
  };

  const handlerAccountType = (e: any) => {
    e.preventDefault();
    const { value } = e.target;
    const accountList = accountState?.selectedAccountList || [];
    const selectedData = find(accountList, (item) => item._id === value);
    setAccountState({
      ...accountState,
      selectedAccount: selectedData,
      selectedAccountNumber: selectedData?.accountNumber || '',
      selectedAccountType: selectedData?.categoryName || 'other',
    });
  };

  const handleInputChange = (e: any) => {
    e.preventDefault();
    const { value } = e.target;
    setAccountState({
      ...accountState,
      selectedAccountNumber: trim(value),
    });
  };
  const onSubmit = (data: FormData, e: any) => {
    e.preventDefault();
    const { accountNumber, amount } = data;
    const submitProps = {
      accountNumber: parseInt(defaultAccountNumber),
      payeeAccountNumber: parseInt(accountNumber),
      amount: parseFloat(amount),
      accountType: type,
    };
    submitDeposit({ variables: { depositInput: submitProps } })
      .then((res) => {
        const { accounts } = res?.data?.deposit;
        if (type === 'Accounts' && !isEmpty(accounts)) {
          dispatch(addAccounts(accounts));
        }
        if (type === 'Cards' && !isEmpty(accounts)) {
          dispatch(addCards(accounts));
        }
        setErrorMessage('');
        setAccountState({
          type: '',
          selectedAccountList: [],
          selectedAccount: {},
          accountDropdown: [],
          selectedAccountNumber: '',
          selectedAccountType: '',
          selectedId,
        });
        setIsCallTransaction(true);
        return setIsSuccess(true);
      })
      .catch((err) => {
        return setErrorMessage(err?.message);
      });
  };

  const getErrorMessage = (message: String | undefined) => {
    return message && <p className="error-message">{message}</p>;
  };

  const validateAmount = (value: any) => {
    const amount = accountState?.selectedAccount.amount || 0;
    const parseValue = parseFloat(value);
    const parseAmount = parseFloat(amount);
    let status = true;
    if (parseValue === 0 || parseAmount < parseValue) {
      status = false;
    }
    return status;
  };

  const closeError = (e: any) => {
    e.preventDefault();
    setErrorMessage('');
    setIsSuccess(false);
  };
  return (
    <form className="row" onSubmit={handleSubmit(onSubmit)}>
      <div className="col-xs-12">
        <Heading HeadingType="h3" ariaLabel="transfer amount">
          Transfer Amount
        </Heading>
      </div>
      <div className="col-xs-12 error-block">
        {errorMessage && (
          <ErrorMessage
            message={errorMessage}
            className="alert"
            onClick={closeError}
          />
        )}
        {isSuccess && (
          <ErrorMessage
            message="Your Transaction have been success."
            className="success"
            onClick={closeError}
          />
        )}
      </div>
      <div className="col-xs-12">
        <SelectDropdown
          id="account"
          name="account"
          label="Select Account"
          options={AccountTypeList}
          placeholder="Select"
          value={accountState.type}
          onChange={(e: any) => handlerAccount(e)}
        />
      </div>
      {!isEmpty(accountState.accountDropdown) && (
        <>
          <div className="col-xs-12">
            <SelectDropdown
              id="accountType"
              name="accountType"
              label={`Select ${accountState?.type} Type`}
              options={accountState.accountDropdown}
              value={accountState?.selectedAccount?._id || 'other'}
              onChange={(e: any) => handlerAccountType(e)}
            />
          </div>
          <div className="col-xs-12">
            <label htmlFor="accountNumber">{`${accountState?.type} Number*`}</label>
            <input
              type="text"
              id="accountNumber"
              value={accountState?.selectedAccountNumber}
              {...register('accountNumber', {
                required: `${accountState?.type} Number is required`,
                maxLength: {
                  value: 20,
                  message: 'You exceeded max length of 50 !',
                },
              })}
              onChange={(e) => handleInputChange(e)}
            />
            {getErrorMessage(errors?.accountNumber?.message)}
          </div>
          <div className="col-xs-12">
            <label htmlFor="amount">Enter Amount*</label>
            <input
              type="number"
              id="amount"
              {...register('amount', {
                required: `Please Enter Amount`,
                validate: (value) =>
                  validateAmount(value) || 'Amount is not valid',
                maxLength: {
                  value: 20,
                  message: 'You exceeded max length of 50 !',
                },
              })}
            />
            {getErrorMessage(errors?.amount?.message)}
          </div>
          <div className="col-xs-12">
            <Button type="submit">Transfer</Button>
          </div>
        </>
      )}
    </form>
  );
};

export default AmountTransfer;
