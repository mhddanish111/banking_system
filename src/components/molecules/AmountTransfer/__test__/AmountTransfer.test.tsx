import '@testing-library/jest-dom';
import { MockedProvider } from '@apollo/client/testing';
import renderer from 'react-test-renderer';
import AmountTransfer from '../AmountTransfer';
import { DEPOSIT } from '../../../../constant/Mutation';

describe('<AmountTransfer />', () => {
  const accounts = {
    _id: '63c1519e9919f0f41a86bd94',
    type: 'Accounts',
    accountType: 'Savings Account',
    categoryName: 'Regular Account',
    accountNumber: 123456789,
    amount: 10000,
    debitCardCSVNumber: 123,
    debitCardNumber: 12312313123,
    expiryMonth: 1,
    expiryYear: 2023,
    limit: 0,
    maxAmount: 0,
    minAmount: 0,
    charge: 0,
    users: [
      {
        _id: '63c1519e9919f0f41a86bd94',
        firstName: 'test',
        lastName: 'test',
      },
    ],
  };
  const props = {
    defaultAccountNumber: 1673894730214,
    selectedId: '63c59b4ae67667f7cbc2ccfe',
    type: 'Accounts',
    state: {
      user: {
        firstName: 'test',
        lastName: 'test',
        userType: 'user',
        userId: '63c1519e9919f0f41a86bd94',
      },
      accounts: [{ ...accounts }],
      cards: [{ ...accounts }],
    },
  };
  const submitProps = {
    accountNumber: 1234567890,
    payeeAccountNumber: 1212121212121212,
    amount: 100,
    accountType: 'Accounts',
  };
  const mocks = [
    {
      request: {
        query: DEPOSIT,
        variables: {
          depositInput: submitProps,
        },
      },
      result: {
        data: {
          deposit: {
            accounts: [{ ...accounts }],
            transactionResult: { _id: '63c1519e9919f0f41a86bd94' },
          },
        },
      },
    },
  ];
  const wrapper = renderer.create(
    <MockedProvider mocks={mocks} addTypename={false}>
      <AmountTransfer {...props} />
    </MockedProvider>,
  ).toJSON;
  it('should render the AmountTransfer Component correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
