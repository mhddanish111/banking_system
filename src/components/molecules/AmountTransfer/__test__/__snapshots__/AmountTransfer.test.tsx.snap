// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AmountTransfer /> should render the AmountTransfer Component correctly 1`] = `[Function]`;
